odbc/src/api.rs

```diff
@@ -709,16 +895,100 @@ pub extern "C" fn SQLGetDiagRec(
 
 #[no_mangle]
 pub extern "C" fn SQLGetDiagRecW(
-    _handle_type: HandleType,
-    _handle: Handle,
-    _record_rumber: SmallInt,
-    _state: *mut WChar,
-    _native_error_ptr: *mut Integer,
-    _message_text: *mut WChar,
-    _buffer_length: SmallInt,
-    _text_length_ptr: *mut SmallInt,
-) -> SqlReturn {
-    unimplemented!()
+    handle_type: HandleType,
+    handle: Handle,
+    rec_number: SmallInt,
+    state: *mut WChar,
+    native_error_ptr: *mut Integer,
+    message_text: *mut WChar,
+    buffer_length: SmallInt,
+    text_length_ptr: *mut SmallInt,
+) -> SqlReturn {
+    if rec_number < 1 || buffer_length < 0 {
+        return SqlReturn::ERROR;
+    }
+    let mongo_handle = handle as *mut MongoHandle;
+    // Make the record number zero-indexed
+    let rec_number = (rec_number - 1) as usize;
+    match handle_type {
+        HandleType::Env => match unsafe { (*mongo_handle).as_env() } {
```

Comment:
nit: Each one of these match branches has quite a bit of code. I think it would be worthwhile to introduce helper functions for each handle type.

Reply:
yeah, similar to what I mentioned in a comment elsewhere, I think this could just be a method on `MongoHandle`